File: Firmware/src/animations/animation_gradient.cpp
#include "animation_gradient.h"
#include "data_set/data_set.h"
#include "data_set/data_animation_bits.h"

namespace Animations
{
	/// <summary>
	/// constructor for rainbow animations
	/// Needs to have an associated preset passed in
	/// </summary>
	AnimationInstanceGradient::AnimationInstanceGradient(const AnimationGradient* preset, const DataSet::AnimationBits* bits)
		: AnimationInstance(preset, bits) {
	}

	/// <summary>
	/// destructor
	/// </summary>
	AnimationInstanceGradient::~AnimationInstanceGradient() {
	}

	/// <summary>
	/// Small helper to return the expected size of the preset data
	/// </summary>
	int AnimationInstanceGradient::animationSize() const {
		return sizeof(AnimationGradient);
	}

	/// <summary>
	/// (re)Initializes the instance to animate leds. This can be called on a reused instance.
	/// </summary>
	void AnimationInstanceGradient::start(int _startTime, uint8_t _remapFace, bool _loop) {
		AnimationInstance::start(_startTime, _remapFace, _loop);
	}

	/// <summary>
	/// Computes the list of LEDs that need to be on, and what their intensities should be.
	/// </summary>
	/// <param name="ms">The animation time (in milliseconds)</param>
	/// <param name="retIndices">the return list of LED indices to fill, max size should be at least 21, the max number of leds</param>
	/// <param name="retColors">the return list of LED color to fill, max size should be at least 21, the max number of leds</param>
	/// <returns>The number of leds/intensities added to the return array</returns>
	int AnimationInstanceGradient::updateLEDs(int ms, int retIndices[], uint32_t retColors[]) {
        int time = ms - startTime;
        auto preset = getPreset();

        // Figure out the color from the gradient
        auto& gradient = animationBits->getRGBTrack(preset->gradientTrackOffset);

        int gradientTime = time * 1000 / preset->duration;
        uint32_t color = gradient.evaluateColor(animationBits, gradientTime);

        // Fill the indices and colors for the anim controller to know how to update leds
        int retCount = 0;
        for (int i = 0; i < 20; ++i) {
            if ((preset->faceMask & (1 << i)) != 0)
            {
                retIndices[retCount] = i;
                retColors[retCount] = color;
                retCount++;
            }
        }
        return retCount;
	}

	/// <summary>
	/// Clear all LEDs controlled by this animation, for instance when the anim gets interrupted.
	/// </summary>
	int AnimationInstanceGradient::stop(int retIndices[]) {
		auto preset = getPreset();
            int retCount = 0;
            for (int i = 0; i < 20; ++i) {
                if ((preset->faceMask & (1 << i)) != 0)
                {
                    retIndices[retCount] = i;
                    retCount++;
                }
            }
            return retCount;
	}

	const AnimationGradient* AnimationInstanceGradient::getPreset() const {
        return static_cast<const AnimationGradient*>(animationPreset);
    }
}

File: Central/Assets/Scripts/Animations/AnimationGradient.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Runtime.InteropServices;

namespace Animations
{
    /// <summary>
    /// Procedural rainbow animation
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class AnimationGradient
        : Animation
    {
        public AnimationType type { get; set; } = AnimationType.Gradient;
        public byte padding_type { get; set; }
        public ushort duration { get; set; }
        public uint faceMask;
        public ushort gradientTrackOffset;
        public ushort gradientPadding;

        public AnimationInstance CreateInstance(DataSet.AnimationBits bits)
        {
            return new AnimationInstanceGradient(this, bits);
        }
    };

    /// <summary>
    /// Procedural on off animation instance data
    /// </summary>
    public class AnimationInstanceGradient
        : AnimationInstance
    {
        public AnimationInstanceGradient(Animation animation, DataSet.AnimationBits bits)
            : base(animation, bits)
        {
        }

        public override int updateLEDs(int ms, int[] retIndices, uint[] retColors)
        {
            int time = ms - startTime;
            var preset = getPreset();

            // Figure out the color from the gradient
            var gradient = animationBits.getRGBTrack(preset.gradientTrackOffset);
            int gradientTime = time * 1000 / preset.duration;
            uint color = gradient.evaluateColor(animationBits, gradientTime);

            // Fill the indices and colors for the anim controller to know how to update leds
            int retCount = 0;
            for (int i = 0; i < 20; ++i)
            {
                if ((preset.faceMask & (1 << i)) != 0)
                {
                    retIndices[retCount] = i;
                    retColors[retCount] = color;
                    retCount++;
                }
            }
            return retCount;
        }

        public override int stop(int[] retIndices)
        {
            var preset = getPreset();
            int retCount = 0;
            for (int i = 0; i < 20; ++i)
            {
                if ((preset.faceMask & (1 << i)) != 0)
                {
                    retIndices[retCount] = i;
                    retCount++;
                }
            }
            return retCount;
        }

        public AnimationGradient getPreset()
        {
            return (AnimationGradient)animationPreset;
        }
    };
}